allennlp/data/dataset_readers/reading_comprehension/util.py

```diff
@@ -63,3 +75,121 @@ def char_span_to_token_span(token_offsets: List[Tuple[int, int]],
     if token_offsets[end_index][1] != character_span[1]:
         error = True
     return (start_index, end_index), error
+
+
+def find_valid_answer_spans(passage_tokens: List[Token],
+                            answer_texts: List[str]) -> List[Tuple[int, int]]:
+    """
+    Finds a list of token spans in ``passage_tokens`` that match the given ``answer_texts``.  This
+    tries to find all spans that would evaluate to correct given the SQuAD and TriviaQA official
+    evaluation scripts, which do some normalization of the input text.
+    """
+    normalized_tokens = [token.text.lower().strip(STRIPPED_CHARACTERS) for token in passage_tokens]
+    # Because there could be many `answer_texts`, we'll do the most expensive pre-processing
+    # step once.  This gives us a map from tokens to the position in the passage they appear.
+    word_positions: Dict[str, List[int]] = defaultdict(list)
+    for i, token in enumerate(normalized_tokens):
+        word_positions[token].append(i)
+    spans = []
+    for answer_text in answer_texts:
+        # For each answer, we'll first find all valid start positions in the passage.  Then
+        # we'll grow each span to the same length as the number of answer tokens, and see if we
+        # have a match.  We're a little tricky as we grow the span, skipping words that are
+        # already pruned from the normalized answer text, and stopping early if we don't match.
+        answer_tokens = answer_text.lower().strip(STRIPPED_CHARACTERS).split()
+        num_answer_tokens = len(answer_tokens)
+        for span_start in word_positions[answer_tokens[0]]:
+            span_end = span_start  # span_end is _inclusive_
+            answer_index = 1
+            while answer_index < num_answer_tokens and span_end + 1 < len(normalized_tokens):
+                token = normalized_tokens[span_end + 1]
+                if answer_tokens[answer_index] == token:
+                    answer_index += 1
+                    span_end += 1
+                elif token in IGNORED_TOKENS:
+                    span_end += 1
+                else:
+                    break
+            if num_answer_tokens == answer_index:
+                spans.append((span_start, span_end))
+    return spans
```

Comment:
You might want to remove duplicate spans here, I would hesitate to assert the its impossible for two different answer strings to "trigger" the same span.

Reply:
The logic that follows actually takes care of this.  It was built for the SQuAD dev set, which can have duplicate annotations, so it picks the most frequent one.  That means we'd bias toward picking spans which work for multiple aliases, which is probably a good thing if it ever happens.  But I'll leave a note here about that, so you don't have to be as familiar with the rest of the code when reading and modifying this piece.